projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,209 @@
+# Summary
+[summary]: #summary
+
+Create a transaction template that can be used to form a transaction by following the instructions.
+
+# Motivation
+[motivation]: #motivation
+
+Some clients require special additions to how the transactions are created (e.g. sending authority to certain addresses after the transaction, allowing the authority address to be from outside the wallet, adding a special output in the transaction, etc.).
+Each of these would be added as an option on the wallet facade and allowing the wallet-lib to create the desired transaction, but each option requires discussion, development time and a new release from both the headless and wallet-lib.
+
+A solution to this is a more flexible API that can receive a transaction template and follow its intructions to form a transaction, if the template is well contructed it can solve all special cases without additional development.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The transaction template will only contemplate version 0x01 transactions which means that we will not create tokens or NFTs with the template for now.
+The transaction template will be a set of instructions on how to create a transaction and how to modify it.
+
+The "user controlled" fields of a transaction are inputs and outputs, which means we can focus on instructions to select inputs and generate outputs.
+
+```ts
+interface TxTemplate {
+  inputs?: InputTemplateInstruction[];
+  outputs?: OutputTemplateInstruction[];
+  actions?: ActionTemplateInstruction[];
+}
+```
+
+## User instructions
+
+### Input instructions
+
+Each input has a `type` argument which will define the other arguments.
+The `type` can be 'utxo', 'raw' or 'action' (defaults to 'utxo' if not present).
+
+The `utxo` inputs will use the storage's `selectUtxos` to find the desired inputs.
+
+- `token`: which token to select, defaults to HTR.
+- `fill`: amount of tokens to select, may require multiple inputs to match.
+- `authority`: 'mint' or 'melt', if present will find the desired authority utxos.
+- `address`: find only utxos of this address.
+- `autochange`: boolean, whether to add a change output if required, defaults to true.
+- `position`: which index to insert the input.
+  - can be `-1` to insert at the end, which is the default behavior.
+  - if the position does not exist, the request will fail.

```

Comment:
We push the other elements, for instance if we have the array [0, 1, 2] and we want to insert a new element 3 on position 1 it would be [0, 3, 1, 2].

Initially I wanted to make position be 'begin' and 'end' only but there are cases where we want to control the position of elements.
In my example I created the transaction, shuffled it and then added the data output at the beggining, but we could add the data output at position 0 then add everything else at position 1, making the position of the data output fixed.

Reply:
It's a good solution. But I think this behavior should be explicit on the description itself, not only inferred by the reader on the example.
Do you agree?

Reply:
It's a good solution. But I think this behavior should be explicit on the description itself to avoid ambiguity, not only inferred by the reader on the example.

Reply:
Closing as the design changed since this comment